src/Sentry/Internal/Http/CachingTransport.cs

```diff
@@ -0,0 +1,271 @@
+            _cacheDirectoryPath = !string.IsNullOrWhiteSpace(options.CacheDirectoryPath)
+                ? _cacheDirectoryPath = options.CacheDirectoryPath
+                : throw new InvalidOperationException("Cache directory is not set.");
+
+            _processingDirectoryPath = Path.Combine(_cacheDirectoryPath, "__processing");
+
+            // Processing directory may already contain some files left from previous session
+            // if the worker has been terminated unexpectedly.
+            // Move everything from that directory back to cache directory.
+            if (Directory.Exists(_processingDirectoryPath))
+            {
+                foreach (var filePath in Directory.EnumerateFiles(_processingDirectoryPath))
+                {
+                    File.Move(
+                        filePath,
+                        Path.Combine(_cacheDirectoryPath, Path.GetFileName(filePath))
+                    );
+                }
+            }
+
+            _worker = Task.Run(async () =>
+            {
+                while (!_workerCts.IsCancellationRequested)
+                {
+                    try
+                    {
+                        await _workerSignal.WaitAsync(_workerCts.Token).ConfigureAwait(false);
+                        await ProcessCacheAsync(_workerCts.Token).ConfigureAwait(false);
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        // Ignore
+                    }
+                    catch (Exception ex)
+                    {
+                        _options.DiagnosticLogger?.LogError(
+                            "Exception in background worker of CachingTransport.",
+                            ex
+                        );
+                    }
+                }
+            });
+        }
+
+        private IEnumerable<string> GetCacheFilePaths()
+        {
+            if (!Directory.Exists(_cacheDirectoryPath))
+            {
+                return Enumerable.Empty<string>();
+            }
+
+            return Directory
+                .EnumerateFiles(_cacheDirectoryPath, $"*.{EnvelopeFileExt}")
+                .OrderBy(f => new FileInfo(f).CreationTimeUtc);
+        }
+
+        public int GetCacheLength() => GetCacheFilePaths().Count();
+
+        // Gets the next cache file and moves it to "processing"
+        private async ValueTask<string?> TryPrepareNextCacheFileAsync(
+            CancellationToken cancellationToken = default)
+        {
+            using var lockClaim = await _cachingDirectoryLock.ClaimAsync(cancellationToken).ConfigureAwait(false);
+
+            var filePath = GetCacheFilePaths().FirstOrDefault();
+            if (string.IsNullOrWhiteSpace(filePath))
+            {
+                return null;
+            }
+
+            var targetFilePath = Path.Combine(_processingDirectoryPath, Path.GetFileName(filePath));
+
+            Directory.CreateDirectory(_processingDirectoryPath);
```

Comment:
Yes, it's not a guarantee, but it's rather cheap if the directory already exists and it's not like we're running in a tight loop anyway. This gives a bit more resiliency because we don't care about previous state.